### PR TITLE
Update OpenAI API to use client-based approach

### PR DESCRIPTION
**Pull request by lgaleana-llm**
This PR updates the OpenAI API usage in the FastAPI application to use the new client-based approach. The changes include updating the `generate_headline` function and modifying the unit tests to mock the new client-based approach.

### Test Plan

pytest / playwright

### CODE REVIEW

**Comment by lgaleana-llm** [comment]
Time to create PR: 224.86 seconds
OPenAI cost: 1.2174

File: test_main.py [hunk 1]
```diff
@@ -19,13 +22,24 @@ def message(self):
...
+def set_openai_api_key():
+    with patch.dict(os.environ, {"OPENAI_API_KEY": "test"}):
+        yield
+
+
 def test_generate_headline():
     text = "This is a sample website text."
     image_url = "http://example.com/image.jpg"
-    with patch("openai.ChatCompletion.create", side_effect=mock_openai_chat_completion_create):
+    with patch("app.openai_utils.get_openai_client") as mock_get_client:
```
Comment (lgaleana) [inline test_main.py]: Can you make this openai mocking into its own method?

Comment (lgaleana-llm) [reply test_main.py]: Certainly! We can refactor the OpenAI mocking into its own method to keep the test code clean and reusable. I'll make the necessary changes and update the PR.

**Comment by lgaleana-llm** [comment]
Time to amend PR: 69.32 secondsOPenAI cost: 0.73858

File: test_main.py [hunk 2]
```diff
@@ -19,13 +22,26 @@ def message(self):
 
     return MockResponse()
 
+@pytest.fixture(autouse=True)
+def set_openai_api_key():
```
Comment (lgaleana) [inline test_main.py]: Can you put these util methods into their own files?

Comment (lgaleana-llm) [reply test_main.py]: Certainly! We can move the utility methods into their own files to keep the test file clean and organized. I'll make the necessary changes and update the PR.

### Plan

1. **Create a New File for Test Utilities:**
   - Create a new file `test_utils.py` to hold the utility methods and fixtures.

2. **Move Utility Methods to `test_utils.py`:**
   - Move the `mock_openai_chat_completion_create`, `set_openai_api_key`, and `mock_openai_client` functions to `test_utils.py`.

3. **Update `test_main.py`:**
   - Import the utility methods from `test_utils.py`.

### Implementation

Let's start by creating the `test_utils.py` file and updating the `test_main.py` file.

### New `test_utils.py`

```python
import pytest
from unittest.mock import patch, MagicMock
import os

def mock_openai_chat_completion_create(*args, **kwargs):
    class MockResponse:
        def __init__(self):
            self.choices = [self]

        @property
        def message(self):
            return {"content": "Mocked Ad Headline"}

    return MockResponse()

@pytest.fixture(autouse=True)
def set_openai_api_key():
    with patch.dict(os.environ, {"OPENAI_API_KEY": "test"}):
        yield

def mock_openai_client():
    mock_client = MagicMock()
    mock_client.chat.completions.create.side_effect = mock_openai_chat_completion_create
    return mock_client
```

### Updated `test_main.py`

```python
import pytest
from fastapi.testclient import TestClient
from unittest.mock import patch
import requests
from bs4 import BeautifulSoup
from app.main import app
from app.openai_utils import generate_headline, get_openai_client
from app.scraping import extract_images
from test_utils import set_openai_api_key, mock_openai_client

client = TestClient(app)

def test_generate_headline():
    text = "This is a sample website text."
    image_url = "http://example.com/image.jpg"
    with patch("app.openai_utils.get_openai_client", return_value=mock_openai_client()):
        headline = generate_headline(text, image_url)
        assert headline == "Mocked Ad Headline"

def test_extract_text_success():
    url = "http://example.com"
    html_content = """
    <html>
        <body>
            <p>Hello, World!</p>
            <img src="http://example.com/image1.jpg" />
            <img src="http://example.com/image2.jpg" />
        </body>
    </html>
    """
    expected_text = "Hello, World!"
    expected_images = ["http://example.com/image1.jpg", "http://example.com/image2.jpg"]
    expected_headlines = ["Mocked Ad Headline", "Mocked Ad Headline"]

    with patch("requests.get") as mock_get, patch("app.openai_utils.get_openai_client", return_value=mock_openai_client()):
        mock_get.return_value.status_code = 200
        mock_get.return_value.content = html_content

        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 200
        assert response.json() == {"text": expected_text, "images": expected_images, "headlines": expected_headlines}

def test_extract_text_invalid_url():
    response = client.get("/extract-text", params={"url": "invalid-url"})
    assert response.status_code == 422

def test_extract_text_request_exception():
    url = "http://example.com"

    with patch("requests.get") as mock_get:
        mock_get.side_effect = requests.RequestException("Request failed")

        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 400
        assert response.json() == {"detail": "Request failed"}

def test_extract_text_bermuda():
    url = "https://thinkingofbermuda.com"
    
    with patch("app.openai_utils.get_openai_client", return_value=mock_openai_client()):
        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 200
        assert "text" in response.json()
        assert "images" in response.json()
        assert "headlines" in response.json()

def test_extract_images():
    html_content = """
    <html>
        <body>
            <img src="http://example.com/image1.jpg" />
            <meta property="og:image" content="http://example.com/image2.jpg" />
            <link rel="image_src" href="http://example.com/image3.jpg" type="image/jpeg" />
            <source srcset="http://example.com/image4.jpg 1x, http://example.com/image5.jpg 2x" />
        </body>
    </html>
    """
    soup = BeautifulSoup(html_content, "html.parser")
    expected_images = [
        "http://example.com/image1.jpg",
        "http://example.com/image2.jpg",
        "http://example.com/image3.jpg",
        "http://example.com/image4.jpg"
    ]
    assert extract_images(soup) == expected_images
```

### Explanation of Changes

1. **New `test_utils.py`:**
   - Created a new file `test_utils.py` to hold the utility methods and fixtures.
   - Moved the `mock_openai_chat_completion_create`, `set_openai_api_key`, and `mock_openai_client` functions to `test_utils.py`.

2. **Updated `test_main.py`:**
   - Imported the utility methods from `test_utils.py`.

I will now update the PR with these changes.

**Comment by lgaleana-llm** [comment]
Time to amend PR: 164.97 secondsOPenAI cost: 1.0439

File: app/openai_utils.py [hunk 3]
```diff
@@ -1,8 +1,14 @@
-import openai
+from openai import OpenAI
+import os
+
+def get_openai_client():
```
Comment (lgaleana) [inline app/openai_utils.py]: I know you created this method to make it easier to mock the openai client. Is there a better way to do this?

Comment (lgaleana-llm) [reply app/openai_utils.py]: Certainly! We can use dependency injection to make it easier to mock the OpenAI client without needing to create a separate method. This approach allows us to pass the OpenAI client as a dependency to the functions that need it, making it more flexible and easier to test.

### Plan

1. **Refactor `generate_headline` to Accept OpenAI Client as a Parameter:**
   - Modify the `generate_headline` function to accept the OpenAI client as a parameter.

2. **Update the Test Cases:**
   - Update the test cases to pass the mocked OpenAI client to the `generate_headline` function.

### Steps

1. **Modify `app/openai_utils.py`:**
   - Refactor the `generate_headline` function to accept the OpenAI client as a parameter.

2. **Update `test_main.py`:**
   - Update the test cases to pass the mocked OpenAI client to the `generate_headline` function.

### Implementation

Let's start by refactoring the `app/openai_utils.py` file and updating the `test_main.py` file.

### Updated `app/openai_utils.py`

```python
from openai import OpenAI
import os

def generate_headline(client, text: str, image_url: str) -> str:
    prompt = f"Generate an ad headline for the following image URL based on the website text:\n\nWebsite Text:\n{text}\n\nImage URL: {image_url}\n\nAd Headline:"
    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": prompt}
        ],
        max_tokens=50
    )
    return response.choices[0].message['content'].strip()
```

### Updated `test_main.py`

```python
import pytest
from fastapi.testclient import TestClient
from unittest.mock import patch
import requests
from bs4 import BeautifulSoup
from app.main import app
from app.openai_utils import generate_headline
from app.scraping import extract_images
from test_utils import set_openai_api_key, mock_openai_client

client = TestClient(app)

def test_generate_headline():
    text = "This is a sample website text."
    image_url = "http://example.com/image.jpg"
    mock_client = mock_openai_client()
    headline = generate_headline(mock_client, text, image_url)
    assert headline == "Mocked Ad Headline"

def test_extract_text_success():
    url = "http://example.com"
    html_content = """
    <html>
        <body>
            <p>Hello, World!</p>
            <img src="http://example.com/image1.jpg" />
            <img src="http://example.com/image2.jpg" />
        </body>
    </html>
    """
    expected_text = "Hello, World!"
    expected_images = ["http://example.com/image1.jpg", "http://example.com/image2.jpg"]
    expected_headlines = ["Mocked Ad Headline", "Mocked Ad Headline"]

    with patch("requests.get") as mock_get, patch("app.openai_utils.OpenAI", return_value=mock_openai_client()):
        mock_get.return_value.status_code = 200
        mock_get.return_value.content = html_content

        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 200
        assert response.json() == {"text": expected_text, "images": expected_images, "headlines": expected_headlines}

def test_extract_text_invalid_url():
    response = client.get("/extract-text", params={"url": "invalid-url"})
    assert response.status_code == 422

def test_extract_text_request_exception():
    url = "http://example.com"

    with patch("requests.get") as mock_get:
        mock_get.side_effect = requests.RequestException("Request failed")

        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 400
        assert response.json() == {"detail": "Request failed"}

def test_extract_text_bermuda():
    url = "https://thinkingofbermuda.com"
    
    with patch("app.openai_utils.OpenAI", return_value=mock_openai_client()):
        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 200
        assert "text" in response.json()
        assert "images" in response.json()
        assert "headlines" in response.json()

def test_extract_images():
    html_content = """
    <html>
        <body>
            <img src="http://example.com/image1.jpg" />
            <meta property="og:image" content="http://example.com/image2.jpg" />
            <link rel="image_src" href="http://example.com/image3.jpg" type="image/jpeg" />
            <source srcset="http://example.com/image4.jpg 1x, http://example.com/image5.jpg 2x" />
        </body>
    </html>
    """
    soup = BeautifulSoup(html_content, "html.parser")
    expected_images = [
        "http://example.com/image1.jpg",
        "http://example.com/image2.jpg",
        "http://example.com/image3.jpg",
        "http://example.com/image4.jpg"
    ]
    assert extract_images(soup) == expected_images
```

### Explanation of Changes

1. **Refactored `generate_headline`:**
   - Modified the `generate_headline` function to accept the OpenAI client as a parameter.

2. **Updated Test Cases:**
   - Updated the test cases to pass the mocked OpenAI client to the `generate_headline` function.

I will now update the PR with these changes.

**Comment by lgaleana-llm** [comment]
Time to amend PR: 888.87 secondsOPenAI cost: 9.860289999999999

File: app/main.py [hunk 4]
```diff
@@ -48,7 +48,10 @@ def extract_text(url: HttpUrl = Query(..., description="The URL to extract text
     # Extract images using the utility method
     images = extract_images(soup)
 
+    # Initialize OpenAI client
+    client = OpenAI(api_key=openai_api_key)
```
Comment (lgaleana) [inline app/main.py:52]: Why not initialize the client inside the module that uses it?

File: test_main.py [hunk 5]
```diff
@@ -3,28 +3,21 @@
...
-
-    return MockResponse()
 
 def test_generate_headline():
     text = "This is a sample website text."
     image_url = "http://example.com/image.jpg"
-    with patch("openai.ChatCompletion.create", side_effect=mock_openai_chat_completion_create):
-        headline = generate_headline(text, image_url)
-        assert headline == "Mocked Ad Headline"
+    mock_client = mock_openai_client()
```
Comment (lgaleana) [inline test_main.py:17]: Put this test in a different file

File: app/main.py (hunk 4 above)
Comment (lgaleana-llm) [reply app/main.py:52]: Initializing the OpenAI client inside the `generate_headline` function in the `app/openai_utils.py` module ensures that the client is encapsulated within the module that directly uses it. This approach promotes better separation of concerns, making the code more modular and easier to maintain. It also simplifies the function signature of `generate_headline`, as it no longer requires the client to be passed as a parameter, reducing the complexity of the function calls throughout the codebase.

**Comment by lgaleana-llm** [comment]
Time to amend PR: 897.26 secondsOPenAI cost: 0.6525000000000001

**Review by lgaleana — APPROVED** [review]

**Pull request merged**